src/mozilla_sec_eia/utils.py

```diff
@@ -78,12 +97,126 @@ def getconn() -> pg8000.dbapi.Connection:
+        if self._metadata_df is None:
+            self._metadata_df = pd.read_sql(
+                "SELECT * FROM sec10k_metadata", self._engine
+            )
+        return self._metadata_df
 
-    def get_file(self, year_quarter: str, path: str) -> storage.Blob:
+    def get_blob(self, year_quarter: str, path: str) -> storage.Blob:
         """Return Blob pointing to file in GCS bucket."""
         return self._bucket.blob(f"sec10k/sec10k-{year_quarter}/{path}")
 
+    def _get_local_path(self, cache_directory: Path, filing: pd.Series) -> Path:
+        """Return path to a filing in local cache based on metadata."""
+        return cache_directory / Path(
+            f"{filing['CIK']}-{filing['year_quarter']}-"
+            f"{filing['Filename'].replace('edgar/data/', '').replace('/', '-')}"
+        )
+
+    def _v_stack_images(self, *images):
+        """Generate composite of all supplied images."""
+        # Get the widest width.
+        width = max(image.width for image in images)
+        # Add up all the heights.
+        height = sum(image.height for image in images)
+        composite = Image.new("RGB", (width, height))
+        # Paste each image below the one before it.
+        y = 0
+        for image in images:
+            composite.paste(image, (0, y))
+            y += image.height
+        return composite
+
+    def get_filing(
+        self,
+        filing: pd.Series,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> Path:
+        """Cache a single filing and return path."""
+        # Create cache directory
+        cache_directory.mkdir(parents=True, exist_ok=True)
+
+        blob = self.get_blob(filing["year_quarter"], filing["Filename"])
+        local_path = self._get_local_path(cache_directory, filing)
+
+        if exists := local_path.exists():
+            blob.update()
+            local_hash = _compute_md5(local_path)
+            remote_hash = blob.md5_hash
+            refresh = remote_hash != local_hash
+
+        if (not exists) or refresh:
+            blob.download_to_filename(local_path)
+
+        return local_path
+
+    def get_filings(
+        self,
+        filing_selection: pd.DataFrame,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> typing.Generator[Path, None, None]:
+        """Caches filings locally for quick access, and returns generator producing paths.
+
+        Args:
+            filing_selection: Pandas dataframe with same schema as metadata df where each row
+                is a filing to return.
+            cache_directory: Path to directory where filings should be cached.
+        """
+        for _, filing in filing_selection.iterrows():
+            yield self.get_filing(filing, cache_directory)
+
+    def get_ex_21_from_filing(
+        self,
+        filing: Path,
+    ) -> Image:
+        """Extract exhibit 21 from raw filing and return PIL Image of table."""
+        ex_21_pat = re.compile(
```

Comment:
This regex seems like it will capture/find all Exhibit 21 documents, but it would be nice if there was some sort of validation function to ensure that it finds a match for every document within a filing selection that contains an Ex. 21 (according to the metadata). Does that make sense?

If it's easier, maybe that can go in later, when we're doing this with larger batches, and not just using a really small subset of documents

Reply:
I've made it so when you create a new `Sec10k` class it will try to extract the exhibit 21 and log a warning if there's a problem. I think that should more or less accomplish this goal, but let me know if there's something I'm missing